text/0000-layer-metadata-files.md

```diff
@@ -0,0 +1,176 @@
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is as follows - 
+
+The current structure of the `<layers>` folder is -
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (buildpack arg `$1` )
+│   ├── build.toml
+│   ├── launch.toml
+│   ├── store.toml
+│   ├── <layer>.toml
+│   └── <layer>
+└── config
+    └── metadata.toml
+```
+
+
+The proposal is to change this structure to - 
+
+
+```
+`/output` (configurable with `CNB_OUTPUT_DIR` or -output-dir) 
+├── group.toml
+├── plan.toml
+├── layers
+│   └── <escaped_buildpack_id> (`CNB_BP_LAYERS_DIR` )
+│       ├── <layer>.toml
+│       └── <layer>
+└── config
+    ├── metadata.toml
+    └── <escaped_buildpack_id> (`CNB_BP_CONFIG_DIR`  )
+        ├── build.toml
+        ├── launch.toml
+        └── store.toml
+```
+
+The top level `<output-dir>` would be the mounted volume that contains the output from buildpacks. The `<layers>` sub-directory would contains the layer contributions by each buildpack under an `<escaped-buildpack-id>` directory. This directory would be passed to the buildpack via the environment variable `CNB_BP_LAYERS_DIR`. Similarly any config files would reside in the `<config>`, under a sub-directory for each buildpack. This directory would be passed to the buildpack via the environment variable `CNB_BP_CONFIG_DIR`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle phases that reference the `<layers>` directory would now take in `<output-dir>` directory instead. The behavior with respect to the various metadata files and the lifecycle would remain the same, however the `analysis`, `restore`, `build` and `export` phases would have to be updated to look for the various layer directories inside the `layers` and `config` sub-folder in the `<output-dir>`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+This will break backwards compatibility for numerous buildpacks that don't use a set of bindings like `libcnb`. This would also result in a one-time loss of layer caches for app images. It would also result in longer path names. We would also have to re-work the export logic to account for the export of the various files which are partially present in the `config` sub-directory and the `layers` sub-directory.
```

Comment:
> This would also result in a one-time loss of layer caches for app images.

For cached layers this might not be true, we could simply restore the layer contents to the new path. This would provide a one-time loss of launch layers however. We should maybe call that out explicitly. E.g. the restorer will not restore layer metata without layer contents if the bp layers dir path has changed, either due to a platform API change (a buildpack API change might be fine, see comment above regarding compatibility strategies). We must also record the platform API that was used on the app image in order to enable this.

Reply:
@jabrown85 Can we add this caveat to the updated draft? If the layers dir changes (either because the default changed between APIs or due to user-provided configuration) we cannot reuse launch layers from the previous image and therefore should not restore the metadata. This is already an issue if the change is user configured but it is an edge case that will suddenly become a very common occurrence (terrible bug that affect everyone) if we change the default layers dir between API versions.

We will probably need an additional RFC to add metadata to `analyzed.toml` that allows us to make that determination... :/